### PR TITLE
Support simple download

### PR DESCRIPTION
**Pull request by charleso**
This is a naive implementation. I'm going to add support for persistent markers in a separate PR.

/cc @nhibberd @tmcgilchrist 

Grab me if you want to go through a quick conceptual clarify of this.

### CODE REVIEW

**Comment by charleso** [comment]
Ping @nhibberd @tmcgilchrist (or anyone else who cares)

File: main/ambiata.hs [hunk 1]
```diff
@@ -13,11 +13,16 @@ import           System.Log.Handler    (setFormatter)
...
+    Upload
+  | Download
   deriving (Eq, Show)
 
 cmd :: Parser Command
-cmd = subparser (command "upload" (info (pure Upload) ( fullDesc <> progDesc "Upload a dataset to Ambiata.")))
+cmd =
+      subparser (command "upload" (info (pure Upload) ( fullDesc <> progDesc "Upload a dataset to Ambiata.")))
+  -- FIX Remove 'internal' when download has been cleaned up
+  <|> subparser (command "download" (info (pure Download) ( fullDesc <> progDesc "Download datasets from Ambiata.")) <> internal)
```
Comment (nhibberd) [inline main/ambiata.hs]: Is this something we want exposes at this point in time ? /unsure

File: doc/download.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+You can use the `ambiata download` command to transfer data securely from
+Ambiata via a download from Amazon S3.
+
+# Required variables
+
+ - `ORGANISATION_ID`.
+   This will be provided by Ambiata.
+   Currently this needs to be set for the `ambiata` CLI,
+   but at some point in the future may be discovered automatically.
```
Comment (tmcgilchrist) [inline doc/download.md]: I'd leave out that second point. Do we want to suggest we may support something in the future?

File: main/ambiata.hs (hunk 1 above)
Comment (charleso) [inline main/ambiata.hs]: > Is this something we want exposes at this point in time

I haven't implemented "marker files" in this PR, so I don't think we want to expose `download` until we're happy with the implementation. That said, we're unlikely to bump the "gold" version in payload, so it's probably not dangerous.

File: src/Ambiata/Cli/Data/Download.hs [hunk 3]
```diff
@@ -29,17 +36,36 @@ newtype DownloadDir = DownloadDir {
...
+  DownloadAccess {
+      downloadTemporaryCred :: TemporaryCreds
+    , downloadPaths :: [ServerFile]
+    } deriving (Eq, Show)
 
 data DownloadEnv =
-  DownloadEnv DownloadDir CommonEnv deriving (Show)
+  DownloadEnv DownloadDir Organisation Endpoint CommonEnv deriving (Show)
 
 newtype ServerFile = ServerFile {
```
Comment (nhibberd) [inline src/Ambiata/Cli/Data/Download.hs]: Might be worth noting the properties that hiding the constructor is trying to uphold, presume it is just that the address is a valid object and not a prefix?

File: src/Ambiata/Cli/Downloads.hs [hunk 4]
```diff
@@ -34,46 +33,35 @@ import           System.IO
...
-  download (withKey (`combineKey` Key name) a) localFile
-  pure . LocalFile $ name
-  where
-    localFile = (unDownloadDir dir) </> T.unpack name
+doDownload :: DownloadDir -> ServerFile -> AWS LocalFile
+doDownload dir a = do
+  let fp = serverFilePath a
+  liftIO $ putStrLn $ "Downloading " <> fp
+  -- NOTE: This will download to a temporary file and then atomically rename once complete
+  download (unServerFile a) (unDownloadDir dir </> fp)
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: This seems wrong. I believe you are taking the basename of an `Address` to create a `ServerFile`, so this download will fail?

File: src/Ambiata/Cli/Data/Download.hs [hunk 5]
```diff
@@ -29,17 +36,36 @@ newtype DownloadDir = DownloadDir {
...
+
+createServerFile :: Address -> Maybe ServerFile
+createServerFile a =
+  fmap (const (ServerFile a)) . basename . key $ a
+
+createServerFileOrFail :: Monad m => Address -> m ServerFile
+createServerFileOrFail a =
+  maybe (fail . T.unpack $ "Invalid ServerFile: " <> addressToText a) return . createServerFile $ a
+
+serverFilePath :: ServerFile -> FilePath
```
Comment (nhibberd) [inline src/Ambiata/Cli/Data/Download.hs]: This does not match `createServerFile` and `unServerFile`. Looks like `createServerFile` should not be calling basename?

Comment (nhibberd) [inline src/Ambiata/Cli/Data/Download.hs]: Sorry I missed the const there, ignore this.

File: src/Ambiata/Cli/Downloads.hs (hunk 4 above)
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Ignore this as well :(

File: test/Test/Ambiata/Cli/Arbitrary.hs [hunk 6]
```diff
@@ -54,11 +55,26 @@ instance Arbitrary TemporaryAccess where
...
 
+instance Arbitrary LocalFile where
+  arbitrary =
+    LocalFile
+      <$> elements weather
+
 instance Arbitrary ServerFile where
-  arbitrary = ServerFile <$> elements cooking
+  arbitrary =
+    (=<<) (createServerFileOrFail) $
```
Comment (nhibberd) [inline test/Test/Ambiata/Cli/Arbitrary.hs]: Probably be nicer in infix :)

File: src/Ambiata/Cli/Json/Download.hs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import           P
+
+
+data ResponseJsonV1  =
+  ResponseJsonV1 {
+    responseJsonV1 :: DownloadAccess
+  } deriving (Eq, Show)
+
+
+instance ToJSON ResponseJsonV1 where
```
Comment (nhibberd) [inline src/Ambiata/Cli/Json/Download.hs]: This should possibly just be in the `test` tree? (Only used for testing i presume)

**Comment by nhibberd** [comment]
Looks good. Will defer documentation specifics to others

**Pull request merged**